languages/csharp/exercises/concept/regular-expressions/.docs/instructions.md

```diff
@@ -0,0 +1,100 @@
+This exercise addresses the parsing of log files.
+
+After a recent security review you have been asked to clean up the organisation's archived log files.
+
+All strings passed to the methods of guaranteed to be non-null and leading and trailing spaces are guaranteed to be removed.
+
+### 1. Identify garbled log lines
+
+You need some idea of how many log lines in your archive do not comply with current standards. You believe that a simple test reveals whether a log line is valid. To be considered valid a line should begin with one of the following strings:
+
+- [TRC]
+- [DBG]
+- [INF]
+- [WRN]
+- [ERR]
+- [FTL]
+
+Implement the `LogParser.IsValidLine()` method to return `false` if a string is not valid otherwise `true`.
+
+```csharp
+var lp = new LogParser();
+lp.IsValidLine("[ERR] A good error here");
+// => true
+lp.IsValidLine("Any old [ERR] text");
+// => false
+lp.IsValidLine("[BOB] Any old text");
+// => false
+```
+
+### 2. Split the log line
+
+A new team has joined the organization and you find their log files are using a strange separator for "fields". Instead of something sensible like a colon ":" they use a string such as "<--->" or "<=>" (because it's prettier) in fact any string that has a first character of "<" and a last character of ">" and any combination of the following "^\*=-" in between.
+
+Implement the `LogParser.SplitLogLIne()` method that takes a line and returns an array of strings each of which contains a field.
+
+```csharp
+var lp = new LogParser();
+lp.IsValidLine("Section 1<===>Section 2<^-^>Section 3");
+// => {"Section 1", "Section 2", "Section 3"}
+```
+
+### 3. Count the number of lines containg a password
+
+It is important to find any passwords included in a file. These will be dealt with automatically but the team needs to know how about passwords occurred in quoted text so that they can be examined manually.
```

Comment:
When is a log line considered a password? It wasn't clear to me that we were looking for the literal text "password", although I did find a reference to it later. I know much of the exercise is rather artificial, but it felt a bit odd to me. I was wondering if we could come up with something else or change it a bit. Maybe this task could be about detecting logs where somehow the company's super secret password was leaked in the logs? I think the thing that might have tripped me up was the "find any passwords" bit, which seems to imply that all passwords are captured.

Some other ideas: what if there was a software malfunction that leaked passwords for login attempts plaintext into the logs and we have to detect the number of occurences here. We could have the password conform to a company policy format or something, where it starts with an uppercase character followed by one or more lowercase characters and ending with a digit or something.

We could have the next task be about replacing the leaked passwords with a "******" placeholder or something like that.

I don't know. This is hard :)

Reply:
> somehow the company's super-secret password was leaked in the logs?

The trouble is you don't need regexes for this.  `String.Contains()` or `String.InexOf` is sufficient.

> the thing that might have tripped me up was the "find any passwords" bit

This is partially adapted an actual security audit I was the victim of.  You care a lot less about passwords that are just sitting around in the file than those that are highlighted as in "password: mysecret" or something.

I have changed the wording to make it clear how to identify a password for our purposes.

> We could have the next task be about replacing the leaked passwords with a "******" placeholder or something like that.

We already are exercising Regex.Replace().  We need something with capture groups.

> We could have the password conform to a company policy format or something, where it starts with an uppercase character followed by one or more lowercase characters and ending with a digit or something.

If my change in wording does do the trick for you then I will compromise on this and adapt task 5 to use this as well.

Let me know.